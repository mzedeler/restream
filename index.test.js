const { actions, sagas } = require('.');
const createSagaMiddleware = require('redux-saga').default;
const {
  all,
  fork,
  takeEvery,
  take,
  put,
} = require('redux-saga/effects');
const fs = require('fs');
const zlib = require('zlib');
const configureStore = require('redux-mock-store').default;

const demo = Boolean(process.env.DEMO);

describe('replug', () => {
  it('can provide piping of streams', (done) => {
    function* mySagas() {
      yield fork(sagas);
      if (demo) {
        yield takeEvery(
          '*',
          function* debug({ type, id }) {
            // eslint-disable-next-line no-console
            yield console.log(`${type} ${id}`);
          }
        );
      }

      // Setting up input stream
      const input = fs.createReadStream(__filename);
      yield put(actions.readableRegister('input', input));
      expect(yield take('*')).toMatchSnapshot();

      // Setting up transform stream
      const gzip = zlib.createGzip();
      yield put(actions.duplexRegister('gzip', gzip));
      expect(yield take('*')).toMatchSnapshot();

      // Setting up writable stream (sink)
      const output = fs.createWriteStream('/tmp/replug-test-output.gz');
      yield put(actions.writableRegister('output', output));
      expect(yield take('*')).toMatchSnapshot();

      // Piping input to transform stream
      yield put(actions.pipe('input', 'gzip'));
      expect(yield take('*')).toMatchSnapshot();
      expect(yield take('*')).toMatchSnapshot();

      // Piping transform stream to output
      yield put(actions.pipe('gzip', 'output'));
      expect(yield all({
        outputPipe: yield take('*'),
        pipeDone: yield take('*'),
        inputEnd: yield take('*'),
        gzipFinish: yield take('*'),
        gzipUnpipe: yield take('*'),
        gzipEnd: yield take('*'),
        outputFinish: yield take('*'),
        outputUnpipe: yield take('*'),
        outputClose: yield take('*'),
      })).toMatchSnapshot();

      done();
    }

    const sagaMiddleware = createSagaMiddleware();
    const mockStore = configureStore([sagaMiddleware]);
    mockStore({});
    sagaMiddleware.run(mySagas);
  });
});
