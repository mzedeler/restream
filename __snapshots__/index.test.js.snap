// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replug can provide piping of streams 1`] = `
Object {
  "id": "input",
  "type": "REPLUG_READABLE_REGISTER_DONE",
}
`;

exports[`replug can provide piping of streams 2`] = `
Object {
  "id": "gzip",
  "type": "REPLUG_READABLE_REGISTER_DONE",
}
`;

exports[`replug can provide piping of streams 3`] = `
Object {
  "id": "output",
  "type": "REPLUG_WRITABLE_REGISTER_DONE",
}
`;

exports[`replug can provide piping of streams 4`] = `
Object {
  "id": "gzip",
  "type": "REPLUG_READABLE_PIPE",
}
`;

exports[`replug can provide piping of streams 5`] = `
Object {
  "readableId": "input",
  "type": "REPLUG_PIPE_DONE",
  "writableId": "gzip",
}
`;

exports[`replug can provide piping of streams 6`] = `
Object {
  "gzipEnd": Object {
    "id": "gzip",
    "type": "REPLUG_READABLE_END",
  },
  "gzipFinish": Object {
    "id": "gzip",
    "type": "REPLUG_READABLE_FINISH",
  },
  "gzipUnpipe": Object {
    "id": "gzip",
    "type": "REPLUG_READABLE_UNPIPE",
  },
  "inputEnd": Object {
    "id": "input",
    "type": "REPLUG_READABLE_END",
  },
  "outputClose": Object {
    "id": "output",
    "type": "REPLUG_WRITABLE_CLOSE",
  },
  "outputFinish": Object {
    "id": "output",
    "type": "REPLUG_WRITABLE_FINISH",
  },
  "outputPipe": Object {
    "id": "output",
    "type": "REPLUG_WRITABLE_PIPE",
  },
  "outputUnpipe": Object {
    "id": "output",
    "type": "REPLUG_WRITABLE_UNPIPE",
  },
  "pipeDone": Object {
    "readableId": "gzip",
    "type": "REPLUG_PIPE_DONE",
    "writableId": "output",
  },
}
`;
